### PR TITLE
[AERIE-1879] Commands with Dynamic Repeating Arguments are Out of Order

### PR DESCRIPTION
**Pull request by goetzrrGit**
* **Tickets addressed:** AERIE-1879 & AERIE-1886
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description

Clipper has special commands with dynamic repeating arguments. 

ex.
`SEQ_VAR_CMD( [dynamic_arg -> n]..., opcode);`

The above command can be written below

```
SEQ_VAR_CMD( dynamic_arg1, opcode);
SEQ_VAR_CMD( dynamic_arg1, dynamic_arg2, dynamic_arg3, opcode);
SEQ_VAR_CMD( dynamic_arg1, dynamic_arg2, dynamic_arg3, dynamic_arg4, dynamic_arg5, opcode);
etc.
```

The command dictionary typescript library generator already handles these special commands, but with Javascript's `Objects` ordering are not guaranteed, we need to preserve the order of these dynamic arguments. Also, the arguments are missing when expanding the sequence.

Here is an example with a new bananantion command - `PACKAGE_BANANA`

```
    // arguments in order 
    PACKAGE_BANANA({
        bundle_name1: "Chiquita",
        number_of_bananas1: 43,
        bundle_name2: "Dole",
        number_of_bananas2: 12,
        lot_number: 1093,
      }),
      
      // arguments are out of order
      PACKAGE_BANANA({
        bundle_name2: "Dole",
        number_of_bananas1: 43,
        bundle_name1: "Chiquita",
        lot_number: 1093,
        number_of_bananas2: 12
      }),

      PACKAGE_BANANA("Chiquita",43,"Dole",12,"Blue",1,1093)

```

When you expand a sequence you should have the arguments for the command in the correct order. Also, the `args` should have the correct values. 

```
   {
      type: 'command',
      stem: 'PACKAGE_BANANA',
      time: { type: 'COMMAND_COMPLETE' },
      args: ['Chiquita', 43, 'Dole', 12, 1093],
      metadata: { simulatedActivityId: simulatedActivityId4 },
    },

    {
      type: 'command',
      stem: 'PACKAGE_BANANA',
      time: { type: 'COMMAND_COMPLETE' },
      args: ['Chiquita', 43, 'Dole', 12, 1093],     <----- Arguments are in the correct order
      metadata: { simulatedActivityId: simulatedActivityId4 },
    },
   
    {
      type: 'command',
      stem: 'PACKAGE_BANANA',
      time: { type: 'COMMAND_COMPLETE' },
      args: ['Chiquita', 43, 'Dole', 12, 'Blue', 1, 1093],  <----- All arguments are listed, we are not missing some arguments
      metadata: { simulatedActivityId: simulatedActivityId4 },
    }

```

## Verification
Ran through this by hand and also used the JEST tester

To run the JEST test:
 * Nuke everything:
* `./gradlew assemble`
* `docker compose -f docker-compose.yml up`
* Navigate` to the command-expansion-server directory in the AERIE repo
* `nvm use 16`
* `npm run test`

**Note:** you might hit the timeout when running the test for the first time. Just rerun the test and it should run a lot faster.

## Documentation
None

## Future work
~~[AERIE-1886](https://jira.jpl.nasa.gov/browse/AERIE-1886) Command Expansion: Commands with Dynamic Repeating Arguments are Dropped~~

I combined both tickets to this PR

### CODE REVIEW

File: command-expansion-server/src/lib/codegen/CommandEDSLPreface.ts
Comment (dyst5422) [inline command-expansion-server/src/lib/codegen/CommandEDSLPreface.ts]: Use for...of instead

File: command-expansion-server/src/lib/codegen/CommandTypeCodegen.ts
Comment (dyst5422) [inline command-expansion-server/src/lib/codegen/CommandTypeCodegen.ts]: No need of a new scope/return here

```
arg => `"${arg.name}${n}"
```

File: command-expansion-server/src/lib/codegen/CommandTypeCodegen.ts
Comment (dyst5422) [inline command-expansion-server/src/lib/codegen/CommandTypeCodegen.ts]: Same as above comment

**Review by dyst5422 — CHANGES_REQUESTED** [review]

File: command-expansion-server/src/lib/codegen/CommandEDSLPreface.ts
Comment (goetzrrGit) [reply command-expansion-server/src/lib/codegen/CommandEDSLPreface.ts]: Thanks for the suggestion

Comment (goetzrrGit) [reply command-expansion-server/src/lib/codegen/CommandEDSLPreface.ts]: Thanks for the suggestion

File: command-expansion-server/src/lib/codegen/CommandTypeCodegen.ts
Comment (goetzrrGit) [reply command-expansion-server/src/lib/codegen/CommandTypeCodegen.ts]: Thanks for the suggestion

Comment (goetzrrGit) [reply command-expansion-server/src/lib/codegen/CommandTypeCodegen.ts]: Thanks for the suggestion

File: command-expansion-server/src/lib/codegen/CommandTypeCodegen.ts
Comment (goetzrrGit) [reply command-expansion-server/src/lib/codegen/CommandTypeCodegen.ts]: Thanks for the suggestion

Comment (goetzrrGit) [reply command-expansion-server/src/lib/codegen/CommandTypeCodegen.ts]: Thanks for the suggestion

**Review by pcrosemurgy — APPROVED** [review]
Nice work Ryan! All tests passing for me locally.

As part of this PR I would strongly suggest working around / fixing the timeout test failure that occurs upon the first run of `npm run test`. I think that it's okay if our command expansion server tests are not fully automated but they should at least be deterministic (in the sense that both the 1st and 2nd runs yield the same results).

Not going to hold this PR up over this so I'll let you decide if you want to tackle that here or not.

File: command-expansion-server/src/lib/codegen/CommandTypeCodegen.ts
Comment (mattdailis) [inline command-expansion-server/src/lib/codegen/CommandTypeCodegen.ts:122]: Double checking - is the `_ARGS_ORDERS` (with the `S` on the end) correct? Elsewhere I seem to recall seeing `_ARGS_ORDER` singular 🤔

Comment (dyst5422) [reply command-expansion-server/src/lib/codegen/CommandTypeCodegen.ts:122]: Concur, it should be without an `S`

Comment (dyst5422) [reply command-expansion-server/src/lib/codegen/CommandTypeCodegen.ts:122]: Actually, I think it is because this is fundamentally different since there are many valid argument orderings. So the `S` designates that difference?

Comment (goetzrrGit) [reply command-expansion-server/src/lib/codegen/CommandTypeCodegen.ts:122]: This data structure is a list of argument order so I made it an 's' for this one. A list of arrays

**Review by dyst5422 — APPROVED** [review]

**Pull request merged**
